From 249cd3605fe38099f04934dbbfcba654c51850d0 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Sun, 25 Sep 2022 07:35:31 -0400
Subject: [PATCH 1/9] Implement --network option

One of the use cases I can think of is for VPNs separate from the host.
Android work profiles are conceptually similar to Distrobox and also
allow a VPN separate from the main profile.

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create               | 26 ++++++++++++++++++++++----
 docs/usage/distrobox-create.md |  2 ++
 man/man1/distrobox-create.1    |  1 +
 man/man1/distrobox.1           |  1 +
 4 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/distrobox-create b/distrobox-create
index add0e94a..91aca79d 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -58,6 +58,8 @@ container_user_gid="$(id -rg)"
 container_user_home="${HOME:-"/"}"
 container_user_name="${USER}"
 container_user_uid="$(id -ru)"
+container_network=""
+container_network_default="host"
 distrobox_sudo_program="sudo"
 dryrun=0
 init=0
@@ -143,6 +145,7 @@ Options:
 				of the same environment.
 	--home/-H:		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume:		additional volumes to add to the container
+	--network:		select a network namespace for the container
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks:		additional commands to execute during container initialization
 	--pre-init-hooks:	additional commands to execute prior to container initialization
@@ -258,6 +261,13 @@ while :; do
 				shift
 			fi
 			;;
+		--network)
+			if [ -n "$2" ]; then
+				container_network="${2}"
+				shift
+				shift
+			fi
+			;;
 		--) # End of all options.
 			shift
 			break
@@ -313,6 +323,10 @@ if [ -z "${container_name}" ]; then
 	container_name="$(basename "${container_image}" | sed -E 's/[:.]/-/g')"
 fi
 
+if [ -z "${container_network}" ]; then
+	container_network="${container_network_default}"
+fi
+
 # We depend on a container manager let's be sure we have it
 # First we use podman, else docker
 case "${container_manager}" in
@@ -411,7 +425,7 @@ generate_command() {
 		--hostname \"${container_name}.$(uname -n)\"
 		--ipc host
 		--name \"${container_name}\"
-		--network host
+		--network ${container_network}
 		--privileged
 		--security-opt label=disable
 		--user root:root"
@@ -530,9 +544,13 @@ generate_command() {
 	# This is the bare minimum to ensure connectivity inside the container.
 	# These files, will then be kept updated by the main loop every 15 seconds.
 	result_command="${result_command}
-		--volume /etc/hosts:/etc/hosts:ro
-		--volume /etc/localtime:/etc/localtime:ro
-		--volume /etc/resolv.conf:/etc/resolv.conf:ro"
+		--volume /etc/localtime:/etc/localtime:ro"
+
+	if [ "${container_network}" = "${container_network_default}" ]; then
+		result_command="${result_command}
+			--volume /etc/hosts:/etc/hosts:ro
+			--volume /etc/resolv.conf:/etc/resolv.conf:ro"
+	fi
 
 	# These flags are not supported by docker, so we use them only if our
 	# container manager is podman.
diff --git a/docs/usage/distrobox-create.md b/docs/usage/distrobox-create.md
index 56cf86dd..8046f677 100644
--- a/docs/usage/distrobox-create.md
+++ b/docs/usage/distrobox-create.md
@@ -27,6 +27,8 @@ graphical apps (X11/Wayland), and audio.
 				of the same environment.
 	--home/-H		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume		additional volumes to add to the container
+	--network		select a network namespace for the container
+
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks		additional commands to execute during container initialization
 	--pre-init-hooks	additional commands to execute prior to container initialization
diff --git a/man/man1/distrobox-create.1 b/man/man1/distrobox-create.1
index ba36f825..8ab6699b 100644
--- a/man/man1/distrobox-create.1
+++ b/man/man1/distrobox-create.1
@@ -49,6 +49,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
+--network       select a network namespace for the container
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization
diff --git a/man/man1/distrobox.1 b/man/man1/distrobox.1
index 4a86117d..89248950 100644
--- a/man/man1/distrobox.1
+++ b/man/man1/distrobox.1
@@ -358,6 +358,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
+--network       select a network namespace for the container
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization

From 54eef74fcb4563a763a0cff93fb293b7958963ee Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Fri, 14 Oct 2022 10:10:22 -0400
Subject: [PATCH 2/9] Enable --init if using --network

Distrobox doesn't work correctly without it

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/distrobox-create b/distrobox-create
index 91aca79d..0f0974cd 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -325,6 +325,9 @@ fi
 
 if [ -z "${container_network}" ]; then
 	container_network="${container_network_default}"
+else
+	printf "\--network does not work without --init. Enabling"
+	init=1
 fi
 
 # We depend on a container manager let's be sure we have it

From 9706996762ff47e65131a0cdd797a1797164ec98 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Fri, 14 Oct 2022 10:11:13 -0400
Subject: [PATCH 3/9] Update --network's help

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create               | 3 ++-
 docs/usage/distrobox-create.md | 3 ++-
 man/man1/distrobox-create.1    | 3 ++-
 man/man1/distrobox.1           | 3 ++-
 4 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/distrobox-create b/distrobox-create
index 0f0974cd..75fbc232 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -145,7 +145,8 @@ Options:
 				of the same environment.
 	--home/-H:		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume:		additional volumes to add to the container
-	--network:		select a network namespace for the container
+	--network		select a custom network for the container, as accepted by
+				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks:		additional commands to execute during container initialization
 	--pre-init-hooks:	additional commands to execute prior to container initialization
diff --git a/docs/usage/distrobox-create.md b/docs/usage/distrobox-create.md
index 8046f677..8261e846 100644
--- a/docs/usage/distrobox-create.md
+++ b/docs/usage/distrobox-create.md
@@ -27,7 +27,8 @@ graphical apps (X11/Wayland), and audio.
 				of the same environment.
 	--home/-H		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume		additional volumes to add to the container
-	--network		select a network namespace for the container
+	--network		select a custom network for the container, as accepted by
+				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
 
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks		additional commands to execute during container initialization
diff --git a/man/man1/distrobox-create.1 b/man/man1/distrobox-create.1
index 8ab6699b..d54e35d7 100644
--- a/man/man1/distrobox-create.1
+++ b/man/man1/distrobox-create.1
@@ -49,7 +49,8 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
---network       select a network namespace for the container
+--network       select a custom network for the container, as accepted by
+                podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization
diff --git a/man/man1/distrobox.1 b/man/man1/distrobox.1
index 89248950..166513f9 100644
--- a/man/man1/distrobox.1
+++ b/man/man1/distrobox.1
@@ -358,7 +358,8 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
---network       select a network namespace for the container
+--network       select a custom network for the container, as accepted by
+                podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization

From d28c0c8a1db39ae6cbe0555fb7e4519e6741a5c1 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Fri, 14 Oct 2022 10:11:30 -0400
Subject: [PATCH 4/9] Don't set --init if --network is the default anyway

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/distrobox-create b/distrobox-create
index 75fbc232..212ac311 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -326,7 +326,7 @@ fi
 
 if [ -z "${container_network}" ]; then
 	container_network="${container_network_default}"
-else
+elif [ "${container_network}" != "${container_network_default}" ]; then
 	printf "\--network does not work without --init. Enabling"
 	init=1
 fi

From d772fd21d7a908732b02ed6ea7ef985b4126b3a1 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Fri, 14 Oct 2022 10:38:30 -0400
Subject: [PATCH 5/9] Add forgotten newlines. Delete stray backslash

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/distrobox-create b/distrobox-create
index 212ac311..00df03ac 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -327,7 +327,7 @@ fi
 if [ -z "${container_network}" ]; then
 	container_network="${container_network_default}"
 elif [ "${container_network}" != "${container_network_default}" ]; then
-	printf "\--network does not work without --init. Enabling"
+	printf "The --network option does not work without --init. Enabling\n\n"
 	init=1
 fi
 

From 83f280189429b968f57ca2b3086bf933e787cab5 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Fri, 14 Oct 2022 10:39:15 -0400
Subject: [PATCH 6/9] Mark --network as experimental and add a warning print
 when it's used

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create               | 3 ++-
 docs/usage/distrobox-create.md | 2 +-
 man/man1/distrobox-create.1    | 2 +-
 man/man1/distrobox.1           | 2 +-
 4 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/distrobox-create b/distrobox-create
index 00df03ac..adc68599 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -145,7 +145,7 @@ Options:
 				of the same environment.
 	--home/-H:		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume:		additional volumes to add to the container
-	--network		select a custom network for the container, as accepted by
+	--network		(EXPERIMENTAL) select a custom network for the container, as accepted by
 				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks:		additional commands to execute during container initialization
@@ -264,6 +264,7 @@ while :; do
 			;;
 		--network)
 			if [ -n "$2" ]; then
+				printf "Warning: --network is currently experimental and may not work in some configurations\n\n"
 				container_network="${2}"
 				shift
 				shift
diff --git a/docs/usage/distrobox-create.md b/docs/usage/distrobox-create.md
index 8261e846..a96ceb77 100644
--- a/docs/usage/distrobox-create.md
+++ b/docs/usage/distrobox-create.md
@@ -27,7 +27,7 @@ graphical apps (X11/Wayland), and audio.
 				of the same environment.
 	--home/-H		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume		additional volumes to add to the container
-	--network		select a custom network for the container, as accepted by
+	--network		(EXPERIMENTAL) select a custom network for the container, as accepted by
 				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
 
 	--additional-flags/-a:	additional flags to pass to the container manager command
diff --git a/man/man1/distrobox-create.1 b/man/man1/distrobox-create.1
index d54e35d7..1abbddee 100644
--- a/man/man1/distrobox-create.1
+++ b/man/man1/distrobox-create.1
@@ -49,7 +49,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
---network       select a custom network for the container, as accepted by
+--network       (EXPERIMENTAL) select a custom network for the container, as accepted by
                 podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
diff --git a/man/man1/distrobox.1 b/man/man1/distrobox.1
index 166513f9..03a1821c 100644
--- a/man/man1/distrobox.1
+++ b/man/man1/distrobox.1
@@ -358,7 +358,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
             of the same environment.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
---network       select a custom network for the container, as accepted by
+--network       (EXPERIMENTAL) select a custom network for the container, as accepted by
                 podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization

From a27367eecf7fac89bfe654353329aaa0d6598f15 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Sat, 15 Oct 2022 11:35:20 -0400
Subject: [PATCH 7/9] Check if a --network exists and exit if not

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/distrobox-create b/distrobox-create
index adc68599..dcfa6fa2 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -617,6 +617,12 @@ if [ "${dryrun}" -ne 0 ]; then
 	exit 0
 fi
 
+# Check if a given --network actually exists.
+if ! ${container_manager} network exists "${container_network}"; then
+	printf "Network \"%s\" does not exist.\nFor a list of networks, run:\n\n%s network ls\n\n" "${container_network}" "${container_manager}"
+	exit 0
+fi
+
 # Check if the container already exists.
 # If it does, notify the user and exit.
 if ${container_manager} inspect --type container "${container_name}" > /dev/null 2>&1; then

From d92499d6c21f5310e530a3beebb55dd8b9d463f6 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Sat, 15 Oct 2022 11:39:15 -0400
Subject: [PATCH 8/9] Make init --network aware. Don't sync resolv.conf and
 hosts if set

--network should work without --init now

Signed-off-by: Cloudwalk <cloudwalk@icculus.org>
---
 distrobox-create               |  8 ++++----
 distrobox-init                 | 22 +++++++++++++++++-----
 docs/usage/distrobox-create.md |  2 +-
 man/man1/distrobox-create.1    |  2 +-
 man/man1/distrobox.1           |  2 +-
 5 files changed, 24 insertions(+), 12 deletions(-)

diff --git a/distrobox-create b/distrobox-create
index dcfa6fa2..c1bf52d8 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -58,6 +58,7 @@ container_user_gid="$(id -rg)"
 container_user_home="${HOME:-"/"}"
 container_user_name="${USER}"
 container_user_uid="$(id -ru)"
+container_custom_networking=0
 container_network=""
 container_network_default="host"
 distrobox_sudo_program="sudo"
@@ -146,7 +147,7 @@ Options:
 	--home/-H:		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume:		additional volumes to add to the container
 	--network		(EXPERIMENTAL) select a custom network for the container, as accepted by
-				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
+				podman or docker's --network option. Useful for per-container split tunneling of VPNs	default: host
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks:		additional commands to execute during container initialization
 	--pre-init-hooks:	additional commands to execute prior to container initialization
@@ -266,6 +267,7 @@ while :; do
 			if [ -n "$2" ]; then
 				printf "Warning: --network is currently experimental and may not work in some configurations\n\n"
 				container_network="${2}"
+				container_custom_networking=1
 				shift
 				shift
 			fi
@@ -327,9 +329,6 @@ fi
 
 if [ -z "${container_network}" ]; then
 	container_network="${container_network_default}"
-elif [ "${container_network}" != "${container_network_default}" ]; then
-	printf "The --network option does not work without --init. Enabling\n\n"
-	init=1
 fi
 
 # We depend on a container manager let's be sure we have it
@@ -590,6 +589,7 @@ generate_command() {
 		--group ${container_user_gid}
 		--home \"${container_user_custom_home:-"${container_user_home}"}\"
 		--init \"${init}\"
+		--network \"${container_custom_networking}\"
 		--pre-init-hooks \"${container_pre_init_hook}\"
 		-- '${container_init_hook}'
 		"
diff --git a/distrobox-init b/distrobox-init
index 98ccca8d..9deba329 100755
--- a/distrobox-init
+++ b/distrobox-init
@@ -31,6 +31,7 @@ init=0
 init_hook=""
 upgrade=0
 pre_init_hook=""
+custom_networking=0
 verbose=0
 version="1.4.1"
 # Print usage to stdout.
@@ -55,6 +56,7 @@ Options:
 	--help/-h:		show this message
 	--init/-I:		whether to use or not init
 	--pre-init-hooks:	commands to execute prior to init
+	--network:		use custom networking
 	--upgrade/-U:		run init in upgrade mode
 	--verbose/-v:		show more verbosity
 	--version/-V:		show version
@@ -124,6 +126,13 @@ while :; do
 				shift
 			fi
 			;;
+		--network)
+			if [ -n "$2" ]; then
+				custom_networking="$2"
+				shift
+				shift
+			fi
+			;;
 		--)
 			shift
 			init_hook=$*
@@ -1006,11 +1015,14 @@ eval ${init_hook}
 if [ "${init}" -eq 0 ]; then
 	printf "container_setup_done\n"
 	# Keepalive loop
-	HOST_WATCH="
-		/etc/hosts
-		/etc/localtime
-		/etc/resolv.conf
-	"
+	HOST_WATCH="/etc/localtime"
+	if [ "${custom_networking}" -eq 0 ]; then
+		HOST_WATCH="
+			${HOST_WATCH}
+			/etc/hosts
+			/etc/resolv.conf
+		"
+	fi
 	# disable verbose logging for this phase.
 	set +x
 	while true; do
diff --git a/docs/usage/distrobox-create.md b/docs/usage/distrobox-create.md
index a96ceb77..9a80ab6e 100644
--- a/docs/usage/distrobox-create.md
+++ b/docs/usage/distrobox-create.md
@@ -28,7 +28,7 @@ graphical apps (X11/Wayland), and audio.
 	--home/-H		select a custom HOME directory for the container. Useful to avoid host's home littering with temp files.
 	--volume		additional volumes to add to the container
 	--network		(EXPERIMENTAL) select a custom network for the container, as accepted by
-				podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)	default: host
+				podman or docker's --network option. Useful for per-container split tunneling of VPNs	default: host
 
 	--additional-flags/-a:	additional flags to pass to the container manager command
 	--init-hooks		additional commands to execute during container initialization
diff --git a/man/man1/distrobox-create.1 b/man/man1/distrobox-create.1
index 1abbddee..71856bab 100644
--- a/man/man1/distrobox-create.1
+++ b/man/man1/distrobox-create.1
@@ -50,7 +50,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
 --network       (EXPERIMENTAL) select a custom network for the container, as accepted by
-                podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
+                podman or docker's --network option. Useful for per-container split tunneling of VPNs  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization
diff --git a/man/man1/distrobox.1 b/man/man1/distrobox.1
index 03a1821c..ae37beee 100644
--- a/man/man1/distrobox.1
+++ b/man/man1/distrobox.1
@@ -359,7 +359,7 @@ usb devices and graphical apps (X11/Wayland), and audio.
 --home/-H       select a custom HOME directory for the container. Useful to avoid host\[aq]s home littering with temp files.
 --volume        additional volumes to add to the container
 --network       (EXPERIMENTAL) select a custom network for the container, as accepted by
-                podman or docker's --network option. Useful for per-container split tunneling of VPNs (implies --init)  default: host
+                podman or docker's --network option. Useful for per-container split tunneling of VPNs  default: host
 --additional-flags/-a:  additional flags to pass to the container manager command
 --init-hooks        additional commands to execute during container initialization
 --pre-init-hooks    additional commands to execute prior to container initialization

From ec655d1db8bad47a500aade8f6b47375f720d7b0 Mon Sep 17 00:00:00 2001
From: Cloudwalk <cloudwalk@icculus.org>
Date: Tue, 1 Nov 2022 09:22:12 -0400
Subject: [PATCH 9/9] Don't check network existence if using "host", "bridge",
 and similar

---
 distrobox-create | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/distrobox-create b/distrobox-create
index c1bf52d8..dc0948de 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -618,10 +618,24 @@ if [ "${dryrun}" -ne 0 ]; then
 fi
 
 # Check if a given --network actually exists.
-if ! ${container_manager} network exists "${container_network}"; then
-	printf "Network \"%s\" does not exist.\nFor a list of networks, run:\n\n%s network ls\n\n" "${container_network}" "${container_manager}"
-	exit 0
-fi
+# Special cases for "host" and friends.
+# TODO: Check if the specified ns:, container:, or slirp4netns: actually exist
+case "${container_network}" in
+	host) ;;
+	bridge) ;;
+	private) ;;
+	slirp4netns*) ;;
+	none) ;;
+	ns:*) ;;
+	container:*) ;;
+
+	*)
+		if ! ${container_manager} network exists "${container_network}"; then
+			printf "Network \"%s\" does not exist.\nFor a list of networks, run:\n\n%s network ls\n\n" "${container_network}" "${container_manager}"
+			exit 0
+		fi
+		;;
+esac
 
 # Check if the container already exists.
 # If it does, notify the user and exit.
